crates/rpc/src/server/validator.rs

```diff
@@ -0,0 +1,148 @@
+/// NOTE: This module contains logic that will eventually be moved to the Validator component
+/// when it is added to this repository.
+use std::collections::BTreeSet;
+
+use miden_objects::Word;
+use miden_objects::account::{AccountId, PartialAccount, StorageMapWitness};
+use miden_objects::asset::AssetWitness;
+use miden_objects::block::{BlockHeader, BlockNumber};
+use miden_objects::transaction::{
+    AccountInputs,
+    ExecutedTransaction,
+    PartialBlockchain,
+    TransactionInputs,
+};
+use miden_objects::vm::FutureMaybeSend;
+use miden_tx::auth::UnreachableAuth;
+use miden_tx::{
+    DataStore,
+    DataStoreError,
+    MastForestStore,
+    TransactionExecutor,
+    TransactionExecutorError,
+    TransactionMastStore,
+};
+
+/// Executes a transaction using the provided transaction inputs.
+pub async fn re_execute_transaction(
+    tx_inputs: TransactionInputs,
+) -> Result<ExecutedTransaction, TransactionExecutorError> {
+    // Create a DataStore from the transaction inputs.
+    let data_store = TransactionInputsDataStore::new(tx_inputs.clone());
+
+    // Execute the transaction.
+    let (account, block_header, _, input_notes, tx_args) = tx_inputs.into_parts();
+    let executor: TransactionExecutor<'_, '_, _, UnreachableAuth> =
+        TransactionExecutor::new(&data_store);
+    executor
+        .execute_transaction(account.id(), block_header.block_num(), input_notes, tx_args)
+        .await
+}
+
+/// A [`DataStore`] implementation that wraps [`TransactionInputs`]
+struct TransactionInputsDataStore {
+    tx_inputs: TransactionInputs,
+    mast_store: TransactionMastStore,
+}
+
+impl TransactionInputsDataStore {
+    fn new(tx_inputs: TransactionInputs) -> Self {
+        let mast_store = TransactionMastStore::new();
+        mast_store.load_account_code(tx_inputs.account().code());
+        Self { tx_inputs, mast_store }
+    }
+}
+
+impl DataStore for TransactionInputsDataStore {
+    fn get_transaction_inputs(
+        &self,
+        account_id: AccountId,
+        _ref_blocks: BTreeSet<BlockNumber>,
+    ) -> impl FutureMaybeSend<Result<(PartialAccount, BlockHeader, PartialBlockchain), DataStoreError>>
+    {
+        async move {
+            if self.tx_inputs.account().id() != account_id {
+                return Err(DataStoreError::AccountNotFound(account_id));
+            }
+
+            Ok((
+                self.tx_inputs.account().clone(),
+                self.tx_inputs.block_header().clone(),
+                self.tx_inputs.blockchain().clone(),
+            ))
+        }
+    }
+
+    fn get_foreign_account_inputs(
+        &self,
+        foreign_account_id: AccountId,
+        _ref_block: BlockNumber,
+    ) -> impl FutureMaybeSend<Result<AccountInputs, DataStoreError>> {
+        async move { Err(DataStoreError::AccountNotFound(foreign_account_id)) }
+    }
```

Comment:
Returning an error is fine for now, but we'll need to figure out how to read `AccountInputs` for foreign accounts from `TransactionInputs` (the data should be in the `advice_inputs` - but it may not be simple to extract).

Let's create an issue for this.